File: tests/testdata/codegen/components-pp/go/exampleComponent.go
package main

import (
	"fmt"

	"github.com/khulnasoft/khulnasoft-random/sdk/v4/go/random"
	"github.com/khulnasoft/khulnasoft/sdk/v3/go/khulnasoft"
	"github.com/khulnasoft/khulnasoft/sdk/v3/go/khulnasoft/config"
)

type DeploymentZonesArgs struct {
	Zone khulnasoft.StringInput
}

type GithubAppArgs struct {
	Id            khulnasoft.StringInput
	KeyBase64     khulnasoft.StringInput
	WebhookSecret khulnasoft.StringInput
}

type ServersArgs struct {
	Name khulnasoft.StringInput
}

type ExampleComponentArgs struct {
	Input           khulnasoft.StringInput
	CidrBlocks      map[string]khulnasoft.StringInput
	GithubApp       *GithubAppArgs
	Servers         []*ServersArgs
	DeploymentZones map[string]*DeploymentZonesArgs
	IpAddress       []khulnasoft.IntInput
}

type ExampleComponent struct {
	khulnasoft.ResourceState
	Result khulnasoft.AnyOutput
}

func NewExampleComponent(
	ctx *khulnasoft.Context,
	name string,
	args *ExampleComponentArgs,
	opts ...khulnasoft.ResourceOption,
) (*ExampleComponent, error) {
	var componentResource ExampleComponent
	err := ctx.RegisterComponentResource("components:index:ExampleComponent", name, &componentResource, opts...)
	if err != nil {
		return nil, err
	}
	password, err := random.NewRandomPassword(ctx, fmt.Sprintf("%s-password", name), &random.RandomPasswordArgs{
		Length:          khulnasoft.Int(16),
		Special:         khulnasoft.Bool(true),
		OverrideSpecial: args.Input,
	}, khulnasoft.Parent(&componentResource))
	if err != nil {
		return nil, err
	}
	_, err = random.NewRandomPassword(ctx, fmt.Sprintf("%s-githubPassword", name), &random.RandomPasswordArgs{
		Length:          khulnasoft.Int(16),
		Special:         khulnasoft.Bool(true),
		OverrideSpecial: args.GithubApp.WebhookSecret,
	}, khulnasoft.Parent(&componentResource))
	if err != nil {
		return nil, err
	}
	// Example of iterating a list of objects
	var serverPasswords []*random.RandomPassword
	for index := 0; index < len(args.Servers); index++ {
		key0 := index
		val0 := index
		__res, err := random.NewRandomPassword(ctx, fmt.Sprintf("%s-serverPasswords-%v", name, key0), &random.RandomPasswordArgs{
			Length:          khulnasoft.Int(16),
			Special:         khulnasoft.Bool(true),
			OverrideSpecial: args.Servers[val0].Name,
		}, khulnasoft.Parent(&componentResource))
		if err != nil {
			return nil, err
		}
		serverPasswords = append(serverPasswords, __res)
	}
	// Example of iterating a map of objects
	var zonePasswords []*random.RandomPassword
	for key0, val0 := range args.DeploymentZones {
		__res, err := random.NewRandomPassword(ctx, fmt.Sprintf("%s-zonePasswords-%v", name, key0), &random.RandomPasswordArgs{
			Length:          khulnasoft.Int(16),
			Special:         khulnasoft.Bool(true),
			OverrideSpecial: khulnasoft.String(val0),
		}, khulnasoft.Parent(&componentResource))
		if err != nil {
			return nil, err
		}
		zonePasswords = append(zonePasswords, __res)
	}
	_, err = NewSimpleComponent(ctx, fmt.Sprintf("%s-simpleComponent", name), nil, khulnasoft.Parent(&componentResource))
	if err != nil {
		return nil, err
	}
	err = ctx.RegisterResourceOutputs(&componentResource, khulnasoft.Map{
		"result": password.Result,
	})
	if err != nil {
		return nil, err
	}
	componentResource.Result = password.Result
	return &componentResource, nil
}

File: tests/testdata/codegen/components-pp/nodejs/exampleComponent.ts
import * as khulnasoft from "@khulnasoft/khulnasoft";
import * as random from "@khulnasoft/random";
import { SimpleComponent } from "./simpleComponent";

interface ExampleComponentArgs {
    /**
     * A simple input
     */
    input: khulnasoft.Input<string>,
    /**
     * The main CIDR blocks for the VPC
     * It is a map of strings
     */
    cidrBlocks: khulnasoft.Input<Record<string, khulnasoft.Input<string>>>,
    /**
     * GitHub app parameters, see your github app. Ensure the key is the base64-encoded `.pem` file (the output of `base64 app.private-key.pem`, not the content of `private-key.pem`).
     */
    githubApp: {
        id?: khulnasoft.Input<string>,
        keyBase64?: khulnasoft.Input<string>,
        webhookSecret?: khulnasoft.Input<string>,
    },
    /**
     * A list of servers
     */
    servers: {
        name?: khulnasoft.Input<string>,
    }[],
    /**
     * A map between for zones
     */
    deploymentZones: Record<string, {
        zone?: khulnasoft.Input<string>,
    }>,
    ipAddress: khulnasoft.Input<number[]>,
}

export class ExampleComponent extends khulnasoft.ComponentResource {
    public result: khulnasoft.Output<string>;
    constructor(name: string, args: ExampleComponentArgs, opts?: khulnasoft.ComponentResourceOptions) {
        super("components:index:ExampleComponent", name, args, opts);
        const password = new random.RandomPassword(`${name}-password`, {
            length: 16,
            special: true,
            overrideSpecial: args.input,
        }, {
            parent: this,
        });

        const githubPassword = new random.RandomPassword(`${name}-githubPassword`, {
            length: 16,
            special: true,
            overrideSpecial: args.githubApp.webhookSecret,
        }, {
            parent: this,
        });

        // Example of iterating a list of objects
        const serverPasswords: random.RandomPassword[] = [];
        for (const range = {value: 0}; range.value < args.servers.length; range.value++) {
            serverPasswords.push(new random.RandomPassword(`${name}-serverPasswords-${range.value}`, {
                length: 16,
                special: true,
                overrideSpecial: args.servers[range.value].name,
            }, {
            parent: this,
        }));
        }

        // Example of iterating a map of objects
        const zonePasswords: random.RandomPassword[] = [];
        for (const range of Object.entries(args.deploymentZones).map(([k, v]) => ({key: k, value: v}))) {
            zonePasswords.push(new random.RandomPassword(`${name}-zonePasswords-${range.key}`, {
                length: 16,
                special: true,
                overrideSpecial: range.value.zone,
            }, {
            parent: this,
        }));
        }

        const simpleComponent = new SimpleComponent(`${name}-simpleComponent`, {
            parent: this,
        });

        this.result = password.result;
        this.registerOutputs({
            result: password.result,
        });
    }
}
